### ISSUE TITLE
flashattentionlayer produces degenerate output + 3.76x slowdown vs documented 2-4x speedup

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1346]
## Summary

`FlashAttentionLayer<T>` is documented as a drop-in replacement for `MultiHeadAttentionLayer<T>` with the same math + a 2-4x speedup. In practice on the **compiled fused-Adam training path** (the default for float + Adam models), `FlashAttentionLayer` produces uniformly distributed output that does not change during training — gradients are silently zeroed at the engine level because `Engine.FlashAttention` lacked a `GraphMode.IsActive` lazy-graph recording branch, while every other differentiable engine op has one.

## Reproducer

HarmonicEngine consumer test `Phase_PAPER_A_PathB_FlashAttention_Sanity` (10KB WikiText-2 byte-LM, dModel=128, L=2, ctx=64, vocab=256, Adam, lr=1e-3, epochs=2):

| arm | top-1 | top-5 | ppl | train wall |
|-----|------:|------:|----:|-----------:|
| baseline MHA       | 15.20% | 43.40% | 254.69 | 17.87s |
| FA (master, pre-fix)|  0.00% | 100.00% | 256.00 | 59.99s (3.76x SLOWER) |
| FA (Tensors fix)    |  0.00% |  92.80% | 256.00 | 48.53s (2.72x slower)  |

top-1 = 0% + top-5 = 100% + ppl = V (256) is the signature of uniform softmax output, meaning the readout's Dense weights never moved off random init. The 3.76x slowdown vs documented 2-4x speedup is the symptom of eager FA running every training step inside a compile lambda that should have replayed a compiled plan.

## Root cause (engine side — fixed by paired Tensors PR)

`CpuEngine.FlashAttention<T>` in `AiDotNet.Tensors/src/AiDotNet.Tensors/Engines/CpuEngine.cs` was the **only differentiable engine op** without a `GraphMode.IsActive` branch — all 219 other recorded ops route through `GraphMode.Current.Record{Unary,Binary,Variadic}` when a compiled training plan is being traced. Pre-fix:

- Lazy-graph trace records every op in `forward(input)` except FA.
- FA computes eagerly and returns a materialised `Tensor<T>` with no `LazySource`.
- Downstream lazy ops (Permute + Reshape in `FlashAttentionLayer<T>.Forward`) capture the FA output as a **frozen compile-time constant**.
- `plan.Step()` replays everything upstream of FA (Q/K/V projections with fresh weights) but the FA node is a static literal — gradients never flow back to the Q/K/V/O projection weights or the output bias.

The pre-existing `RecordIfActive("FlashAttention", ...)` in eager-tape mode is unrelated and works correctly — the layer-level isolated training tests pass. The bug is specific to the lazy-graph / compiled-training-plan path.

## Pull requests

- **AiDotNet.Tensors**: [ooples/AiDotNet.Tensors#NEW](https://github.com/ooples/AiDotNet.Tensors/pull/new/fix/flashattention-degenerate-backward) — `fix: flashattention engine.flashattention missing graphmode branch`. Adds a `GraphMode.IsActive` branch to `CpuEngine.FlashAttention<T>` mirroring LayerNorm's recompute-mean-on-replay pattern. Adds regression test `GraphModeRecordingAuditTests.FlashAttention_UnderGraphMode_RecordsLazyNode` (fails pre-fix, passes post-fix).

- **AiDotNet**: [ooples/AiDotNet#NEW](https://github.com/ooples/AiDotNet/pull/new/fix/flashattention-layer-degenerate-output) — `test: flashattentionlayer gradient flow + train regression suite (3 tests)`. Tape-based gradient-flow + training regression tests for `FlashAttentionLayer<T>`.

## Remaining work

The Tensors-side fix removed the engine-level gap (lazy node now recorded with non-null `LazySource`) but the HarmonicEngine consumer test still shows degenerate output (top-1 = 0%, top-5 = 92.80%) with only a partial slowdown improvement (3.76x → 2.72x). There is at least one additional gap somewhere in the compile/replay path between the engine fix and the AiDotNet `CompiledTapeTrainingStep.TryStepWithFusedOptimizer` consumer. Candidate areas for follow-up:

- `LazyGraphCompiler` and `CompiledTrainingPlan.ConfigureOptimizer` may need to extend their op-handling to include the `FlashAttention` lazy node type in the backward-graph walker.
- `BackwardFunctions<T>.FlashAttentionBackward` may need a compile-time backward-emitter equivalent to the eager-tape registration.

This issue tracks both the closed engine-side gap and the remaining consumer-side investigation.

## Repro to confirm fix in future

```csharp
[Fact] public void Repro()
{
    var engine = new CpuEngine();
    var q = Tensor<float>.CreateRandom([1, 2, 4, 8]);
    var k = Tensor<float>.CreateRandom([1, 2, 4, 8]);
    var v = Tensor<float>.CreateRandom([1, 2, 4, 8]);
    using var scope = GraphMode.Enable();
    var output = engine.FlashAttention<float>(q, k, v, scale: null, isCausal: false, out _);
    Assert.NotNull(output.LazySource); // pre-fix: null. post-fix: non-null.
}
```

## Consumer ticket

HarmonicEngine PathB FA sanity test: `tests/Learning/Phase_PAPER_A_PathB_FlashAttention_Sanity.cs`. Branch `research/ledger-overhaul-and-l4-reification`.

**Comment by ooples** [comment]
PRs filed:
- AiDotNet.Tensors engine fix: https://github.com/ooples/AiDotNet.Tensors/pull/362
- AiDotNet test additions: https://github.com/ooples/AiDotNet/pull/1347

**Comment by ooples** [comment]
## Closing summary

The **AiDotNet-side gradient-flow bug from this issue is fully fixed** by the merged engine PR [AiDotNet.Tensors#362](https://github.com/ooples/AiDotNet.Tensors/pull/362) (shipped in NuGet 0.81.3 and consumed by AiDotNet master @ 68f5c69). `Engine.FlashAttention<T>` now records a `LazyNodeType.Custom` node with `BackwardFunctions<T>.FlashAttentionBackward` under `GraphMode.IsActive`, so gradient flow through FlashAttention reaches Q/K/V/O projection weights on the compiled fused-Adam training path.

Verified:
- `FlashAttentionLayer_TrainViaFusedCompiledAdam_EngagesFusedPath` (new regression test, added to the AiDotNet diagnostic branch) — **passes**: a Train call on a Transformer containing FA increments `CompiledTapeTrainingStep<float>.GetFusedStepCount()`.
- The 7 existing FlashAttentionLayer tape-based gradient-flow tests in `AttentionLayersIntegrationTests` all pass on master.

### Consumer-side gap remains in AiDotNet.Tensors — tracked at #396

The "PathB still shows degenerate output despite engine fix" symptom this issue described is **not FA-specific** — it's a general bug in the Tensors `CompiledTrainingPlan` loss-readout path. Diagnosed via 5 controlled bisection experiments (full evidence in [ooples/AiDotNet.Tensors#396](https://github.com/ooples/AiDotNet.Tensors/issues/396)):

| # | Setup | Result |
|---|---|---|
| 1 | Tensors-level direct CCE chain via `CompileTraining + Step()` | correct loss |
| 2 | Same + `ConfigureOptimizer(Adam)` | correct loss |
| 3 | Tensors-level 2-param model-like (matmul+add+CCE) with negative logits | NaN (propagated correctly) |
| 4 | Same as #3 with positive logits | correct loss |
| 5 | **AiDotNet** path (Dense+LayerNorm+SeqSlice+Dense IdentityActivation + CCE + fused Adam) | **literal 0** (silent failure) |

When `CategoricalCrossEntropyLoss` is fed raw logits (Dense layer with IdentityActivation), `log(negative_logit + eps)` produces NaN inside the loss chain. The NaN should propagate through ReduceSum → ReduceMean → TensorNegate → `_lossOutput[0]`. Instead, the AiDotNet fused-Adam path returns literal float zero. Consumer sees "loss is 0, training converged" while gradients are corrupted. The 3.76× perf slowdown originally reported in this issue is likely the same underlying class of problem (silent corruption in the compile/replay path).

The actual fix lives in `AiDotNet.Tensors/Engines/Compilation/CompiledTrainingPlan.cs` — either in the forward-action pruning passes (loss node may be considered "unused" and dropped), parameter-count-dependent fusion logic, or in tensor aliasing during lazy realization. The bug is parameter-count-dependent: minimal traces with just `[W, b]` correctly propagate; full layer-stack traces produce 0.

### Action items

- Closing this issue: the engine-fix side is done; consumer-side fix has its own home at AiDotNet.Tensors#396.
- The diagnostic test file (`FlashAttentionFusedCompiledTrainingIssue1346Tests.cs` on branch `fix/1346-flashattention-compile-backward`) can be cherry-picked or re-landed once #396 ships, to gate that the consumer-side fix actually reaches AiDotNet.
- Until #396 lands, **consumers using `CategoricalCrossEntropyLoss` with raw-logit Dense output + float + Adam** should switch to `CrossEntropyWithLogitsLoss<float>` (which applies log_softmax internally and stays numerically stable). This is the proper fix for the user-facing degeneration whether or not the readout bug is fixed.
